Define Minetest's direction

rubenwardy opened the pull request:
This PR aims to define the direction, whilst in a fairly unrestrictive way. This should maybe go on the dev wiki instead of the MT repo, but putting it here is easier to get approval

rubenwardy commented:
Things also worth mentioning:

* Something addressing why the discord patch was rejected. Maybe relating to free vs proprietary and also the low user benefit of it
* More example gameplay targets

paramat commented on doc/direction.md:
```diff
@@ -0,0 +1,52 @@
+### 0.2. Definitions
+
+* `Minetest` is the game engine.
+* `Minetest Game` is the default game which comes with the engine.
```
Yes i think so, as it is still the only bundled game. It needs mentioning as providing some good code examples, although in a poor mod structure.

paramat commented on doc/direction.md:
```diff
@@ -0,0 +1,52 @@
+It's important to limit the scope of the engine to something which is managable.
+
+* Games where players are in 1st or 3rd person.
+* Games with roughly meter cubed cubes.
```
I agree.

paramat commented on doc/direction.md:
```diff
@@ -0,0 +1,52 @@
+
+Please note that Minetest is developed by a community, and not everyone in the community
+holds the same views as to what Minetest should be. Therefore, defining a common direction is
+very difficult.
```
I think most of this text is obvious and doesn't need to be stated.

SmallJoker replied:
Just drop the last message. I think some people should be reminded that not everybody shares their goals.

paramat commented on doc/direction.md:
```diff
@@ -0,0 +1,52 @@
+
+### 0.3. Other documents
+
+Please read celeron55's document: http://c55.me/blog/?p=1491
```
This roadmap in the forum is more detailed and a continuation of that blog post https://forum.minetest.net/viewtopic.php?t=9177

paramat replied:
Also see celeron55's posts here https://github.com/minetest/minetest/issues/3476 for more roadmap.

rubenwardy replied:
I disagree, the article contains more things relevant to direction. The thread is a roadmap, which is distinct to direction. I'll add it anyway

paramat replied:
Yes, sorry, it's mostly specific things to implement, but there is also some important 'direction' material in that roadmap if you read through it. So, perhaps this document can do the useful job of extracting the direction stuff from this roadmap?
```
Focus on these kinds of gameplay

    Voxel-based mechanisms that work well in multiplayer environment
        This works well together with the moddability and people have been playing with mesecons, pipes and whatnot for a long time already.
    Defending against computer-controlled enemies
        Defending instead of attacking puts much better use to the aspect of building things... at least until VAEs exist!
        PvP is generally very hard to balance without making it boring
        Many types of this require a world that is more fully in an active state than how Minetest currently works.
            Currently it is assumed that every game wants to maximize world size and player count by taking away from other things, which is not right.
            It should be possible for a mod or game to pick the tradeoff between world size, active entity count, individual entity performance and the resulting maximum player count.
```
```
Some general guidelines

    Networking: The low-level protocol implementation is bloated and not ideal. See extra information in spoiler below. Packet handling, serialization and cross-version compatibility should be made more easily reviewable and more noob-proof.
    Client UI: Keep it generic and customizable so that it works for any game and any server.
    Pull requests: If you test a pull request, always comment on it saying that you did so. The amount of testers is a very important factor in deciding whether to merge or not.
    Maintain fork-friendliness in everything including versioning of things. See http://irc.minetest.ru/minetest-dev/201 ... #i_3654822
    If you want to focus on exploration of the infinite world, you have to put way more effort into emergent generation and active systems in the world in order to actually have infinite things in there. With almost any fixed selection of static things, it's incredibly bland. For the king of this stuff, see Dwarf Fortress.
    Don't mind the lesser visuals and take full advantage of them in how they make creating unimagined things easier. There are enough games in this world that focus on visuals by taking away from everything else, or by making everything else harder to do.
    If you make game content, read this: viewtopic.php?p=139470#p139470
```

paramat replied:
.. and also extract the direction-related stuff from #3476

paramat commented:
> Something addressing why the discord patch was rejected.

I don't think that needs to be mentioned, doing so just gives an impresion of weakness, trying too hard to apologise to discord fans, as well as overstating the importance of discord.

rubenwardy commented:
> I don't think that needs to be mentioned, doing so gives an impression of weakness, trying too hard to defend ourselves and apologise to discord fans, as well as overstating the importance of discord.

I didn't mean directly mention discord, but rather say something about the scope of changes and whether we find proprietary platforms acceptable to support

SmallJoker commented on doc/direction.md:
```diff
@@ -0,0 +1,52 @@
+## 1. Mission Statement
+
+Minetest provides both a game engine, which allows the creation of voxel games using Lua, and
+a content distribution platform. Players can connect to servers with their own game play, and
```
The master server is also a service.

SmallJoker commented on doc/direction.md:
```diff
@@ -0,0 +1,52 @@
+* **Stability and performance**, especially on older devices.
+* **Keep the engine lightweight**. Support a large amount of gameplay by pushing code and functionality out to Lua,
+  rather than having a complex core. Keeping the core as small as possible makes it
+  easier to test and more flexible.
```
I'd rather head into the direction of suggesting to make features as generic as possible to offer many functionalities for less code.

SmallJoker commented on doc/direction.md:
```diff
@@ -0,0 +1,52 @@
+  If everything is either disableable or suitable for such, then it gives much more freedom
+  for people to experiment and create interesting things, even in ground-based worlds.
+* **Make polishing easy**. It shouldn't be made unnecessarily hard to make polished content
+  even while most content isn’t polished.
```
Would you please be so nice to explain me what you mean by this sentence?

rubenwardy replied:
it's from c55's roadmap

DS-Minetest replied:
Haha, this sounds like nobody knows what it means, but it has to be carried on for completeness. Maybe there's someone who understands this and who would blame you if you dropped this.

c55 uses "mustn't", "not shouldn't".

I think, what this wants to say is that there's much content (eg. mods code) that uses deprecated stuff, has bad formatting, wasn't updated for a long time and is buggy, but creating new content (eg. a new mod) should be easy to do correct, meaning that you should not need examples to make your own code or that you can easily read what features are deprecated.

TumeniNodes replied:
Sometimes, I picture celeron55 poppin onto the web and thinking... "OMFG, it's _still_ going!?"

The overall "direction" seems to be this way ------>
No, wait.... <-------- that way....
wait... give me a second...

Over the next year, MTG just needs to add some simple mobs (a few peaceful such as animals, and a couple aggressive) whether the pathfinding is perfected or not yet... and then add SmallJokers new bows mod, and players will be happy as a pig in...
...mud

This will keep the hordes at bay for a little while, while everyone figures out wtf to do, to put a solid direction/goals to focus on, whether it be an entirely new game to be packaged or what ever...
Even if it is similar to the current MTG but with a lot of current changes without all the patches & bondo covering the dents, causing bloat.

paramat replied:
> Would you please be so nice to explain me what you mean by this sentence?

I think it admits that most MT content isn't very 'polished', but warns against using that fact to justify making polished content excessively difficult to create.
Or maybe, it explains that it is not hypocritical to make polished content possible while most MT content isn't polished.
Makes sense to me.

paramat commented:
> I didn't mean directly mention discord, but rather say something about the scope of changes and whether we find proprietary platforms acceptable to support

Yes, sorry, i misunderstood your intention as a kind of apology or defence of the decision. Disregard =)

paramat commented:
> This should maybe go on the dev wiki instead of the MT repo

I think this should be in the repo, whether it is in the wiki or not. So that it is quickly and easily accessible.

appgurueu commented:
Minetest should stop using Irrlich. First, it doesn't do much for the rendering, and what it does is sometimes buggy. Second, it's outdated and doesn't even support a modern GLSL version (I stumbled upon this problem working on better shaders). Also, you might want to add "2nd rendering stage / better graphics" to the list.

paramat commented:
I think that celeron55's direction-related writings should be reproduced in full (that is, the parts of the texts relevant to direction) and unedited in this document. Not interpreted and summarised, as doing so inevitably alters them, and therefore also alters the officially stated direction of Minetest, based on whoever is doing the interpreting / summarising.
I think the officially stated direction of Minetest is so important it should be in celeron55's words, i don't trust any core dev, including myself, to rewrite or summarise it themselves.

The original texts are then preserved in case they are lost from the forum, or lost from celeron55's blog, or lost from the GitHub issue.

Celeron55's writings have a relaxed, playful style with a sense of humour. This 'relaxed, playful, humourous' attitude is an important part of the direction and philosophy of Minetest (and is something that attracts me to Minetest). Rewriting his texts removes this.

This document can link to the full texts, and then also reproduce the direction-related parts of those texts.

rubenwardy commented:
The majority of celeron55's texts really aren't that useful, and he's not the *sole* defining person for the direction of Minetest. He hasn't been involved for a few years. They also contain typos and such, which should be edited.

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,65 @@
+
+* `Minetest` is the game engine.
+* `Minetest Game` is the default game which comes with the engine.
+* The `Minetest Project` refers to the combination of the above, the development team, and supporting tools.
```
I think that "supporting tools" should be defined too. Media server and mod repository are important enough to be mentioned.

AKryukov92 commented:
I think defining directions and design principles is very important job.
Attitude "lets use old writings of some important man, who has long gone" sounds like religion, not like good design phylosophy. I mean no offence to celeron55, but principles have to be discussed and questioned.

> This approach will also greatly reduce arguments between core devs about the content of the stated 'direction of Minetest'.

Reducing arguments in this manner is _Argumentum ad verecundiam_. It will only sweep conflict under the carpet and pretend that there is no conflict.
I think arguments should be discussed and concensus should be found.

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,65 @@
+## 1. Mission Statement
+
+Minetest provides both a game engine, which allows the creation of voxel games using Lua, and
+a content distribution platform. Players can connect to servers with their own gameplay, and
```
> their own gameplay

I think it should be reprased to emphasise the fact that "servers has own gameplay" not that "players has own gameplay".

ClobberXD replied:
I agree, the wording is ambiguous at present.

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,65 @@
+  easier to test and more flexible.
+* **Make polishing easy**. It shouldn't be made unnecessarily hard to make polished content
+  even while most content isn’t polished.
+* **Prefer simplicity**. Try to be simple in both API and features.
```
I think that word "simple" should be elaborated. Simple for who? For example mapgens might be simple for paramat, but very complex for me. Does mapgen algorithm contradicts to "simplicity" aim?
Software is complex by nature. Implementation of interesting feature might require sophisticated technical details.

I think it is better to write something like "API should be documented. Each function should be provided with examples on wiki. Each development feature should be added for one purpose so modders could combine them to make something complex."

ClobberXD replied:
> For example mapgens might be simple for paramat, but very complex for humans

FTFY :)

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,65 @@
+
+The engine is designed for games which fit these rough limitations:
+
+* Players are in 1st or 3rd person.
```
What about isometric view? What about RTS-style kind of gameplay?

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,65 @@
+The engine is designed for games which fit these rough limitations:
+
+* Players are in 1st or 3rd person.
+* Roughly meter sized cubes.
```
Does it even matter?
Cubes which feel like 0.5 meter size are prohibited to implement?
Large player model, which will have dimensions like 2x2x3 is prohibited?

SmallJoker replied:
No, cubes as in `drawtype = "normal"` are engine-defined. One distance in the grid is one meter per definition.

AKryukov92 replied:
Still, some might want to implement "huge world". In such world, each cube will be equal to 1 inch and nothing will actually change from engine point of view.

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,65 @@
+
+---------------------------
+
+## 1. Mission Statement
```
I have seen some guys on the forums, who use MT for educational purposes. I think it worth mentioning in mission too.

benrob0329 replied:
MT For Education isn't a "goal", it's more of a side-effect due to the nature of MT. Facilitating any such "mucking about" is somewhat of the goal of MT, and specific examples of it don't necessarily need to be mentioned.

paramat replied:
I agree with benrob0329, use in education is not a primary part of the mission.

rubenwardy replied:
Education is an interesting use case but it isn't necessary to mention it

lhofhansl commented:
I think someone who knows that area should improve MT's graphics. Reflective water, real shading, ray traced shadows, etc, etc.

Now inevitably some will object with something "But will nobody think about the poor old hardware?" IMHO, optimizing MT on hardware 5 or more years old while the rest of the world moves on is rather pointless.
The use of the engine can remain just as simple as now, while it can default to more visually pleasing output.

benrob0329 commented:
Someone needs to fix our render pipeline so that it doesn't suck first. Shaders, particles, entities, none of these things are difficult to render, yet MT starts to crawl once they get involved. The whole split pipeline doesn't help either, as all graphical features require twice the effort.

Calinou commented:
> IMHO, optimizing MT on hardware 5 or more years old while the rest of the world moves on is rather pointless.

The average desktop PC in use is more than 5 years old :slightly_smiling_face:

rubenwardy commented:
The priority should be on fixing our graphics code to be less broken and performant, before adding more graphics features. Whilst I would love more fancy graphics, both celeron55 and the general consensus I see is that gameplay and flexibility is more important than graphics

rubenwardy commented:
I'd love it if Minetest could eventually be configured to look like this: https://youtube.com/watch?v=Rkydse1LVX0

Even if it's a shallow fork to add new shaders

TumeniNodes commented:
> I'd love it if Minetest could eventually be configured to look like this: https://youtube.com/watch?v=Rkydse1LVX0
> 
> Even if it's a shallow fork to add new shaders

All of that beauty and splendor....
and then there's "that grass though" xD
It looks so out of place in the video.

Wuzzy2 commented:
I think this document is mostly OK. Two complaints, however:

> * Roughly meter sized cubes.

I call bullshit on that one. This one is the scope for games, not the engine.
The engine should not know about such arcane concepts such as the “meter”. This is a direct contradiction to the goal of being generic.

What about a game in which you play as a small ant? It makes sense to assume each cube is only a cubic millimeter in size. And you have nodes like dust particles, and acorns would be absolutely massive, spanning multiple nodes, etc. And it would all perfectly work with no special cases required.
Basically, length units are arbitrary, it's the games which assign meaning to them.
Therefore, the assumption that every node is 1 cubic meter is bullshit.

While it's true the cubic meter assumption is the *de facto* default, its in no ways safe to assume its true for all games.

---

I also note the lack of a proper definition of a “voxel game”.

rubenwardy commented:
The engine needs to make assumptions about the types of games created. One of the assumptions made is that the world is made up of a voxel grid, for example. The engine doesn't aim to be generic, that's something left to Godot and Unity

The engine doesn't support very small nodes, and shouldn't. It also won't work with massive nodes. "roughly" was intended to mean an order of a magnitude - ie: 0.5-5

benrob0329 commented:
Perhaps describing the engine as supporting nodes which are 1 unit^3 in size, one unit being the size of a "normal" drawtype node on the world grid (or `-0.5,0.5` ^3)

Wuzzy2 commented:
Rubenwardy, you don't get it …

I did not mean that Minetest is meant to be 100% generic. Obviously it's for voxel games, I don't deny that.

I just deny that each node is automatically 1 cubic meter. How did you manage to just dismiss my ant example?

The trick is very simple: In my ant game example, the size of the nodes does !!!!!NOT!!!!! actually change. I'm !!!!!NOT!!!!! saying the engine must shrink all nodes by a factor of 1000.
Actually, nothing at all in the engine changes. It's just the game's textures, environment, models, etc. are designed in such a way that makes you feel very small.
In this game it would be silly to say each node is 1 cubic meter when you have grass that is towering over 100 nodes high (because you're an ant).

I think this whole “meter” convention is silly. Doesn't the engine just see raw, unitless numbers? Basically every game could just use the node size, and *pretend* them to be of any size they want, depending on what kind of environment they want.

Just drop this cubic meter nonsense. It's implying that this ant game would be impossible, which is false.

Maybe what you *actually* wanted to say is something like this:

* Support node sizes roughly from half to up to 5 times their default size

AKryukov92 commented:
> Support node sizes roughly from half to up to 5 times their default size

With the restriction that all nodes in world have same size. It is not possible to make some nodes to be 1x1x1 while others are 2x2x2 *simultaneously* in the same world. Am I right?

rubenwardy commented:
It appears that I did misunderstand you. This is exactly what I intended to say

> Support node sizes roughly from half to up to 5 times their default size

The idea is to rule out variable level of detail voxel grids, and also voxel grids which are either much smaller or much larger than the default size.

AKryukov92 commented:
I'm still not sure that "default size" is. May be we should express it in comparison to player?
Something like
> Size of the player is expected to be no less than 0.5 of block and not more than 3 nodes in each dimension

Regarding definition of "voxel game":
* Things visually can be smaller than full block, but they still take place of the full block. For example torches and halfblocks.
* It is not possible to put two things in the same node. If someone want it for gameplay purpose, then he should use entities or create special block which will look like combination of small things. For example plants in pots and items in frames.
* It is not possible to put one thing in the two nodes. If someone want it, then he should use special blocks, which is part of whole. For example doors.

Though, I'm not sure that last two points should be in "direction". They looks like guideline.

benrob0329 commented:
Oversized nodes are very much supported (MTG's doors are oversized after all) but only up to 3x3 nodes.

raymoo commented:
>Size of the player is expected to be no less than 0.5 of block and not more than 3 nodes in each dimension

I don't really like this, it rules out being able to have different players have vastly different sizes (e.g. mouse vs person).

rubenwardy commented:
I plan on removing the random sizes and just say "may be any size, but optimised for near the default"

AKryukov92 commented:
How much is "default"? It still implies that there is some other unit of size to compare.

What about defining size of the node as [Base unit](https://en.wikipedia.org/wiki/Base_unit_(measurement)) - smallest piece of environment available for interaction. Node size is a standard to measure distance and length in the game and has no fixed real-world derivations. Each game can define own granularity of the world.

rubenwardy commented:
The default size is the default size

SmallJoker commented:
> What about defining size of the node as Base unit - smallest piece of environment available for interaction.

And how would you describe entity positions and dimensions? There's no smallest piece for those (except floating point epsilon).
The "base unit" here would rather be the distance from a whole coordinate to the next adjacent whole coordinate position. I'd rather not rely on defaults for simplicity here. The coordinate system does not change, but defaults could.

Wuzzy2 commented:
Well, in that case, the word “meter” should be only used as a shorthand for “one default node length”, purely for devs and in debugging. It should be made clear that this is meant only as a developer shorthand.

It should not be implied that's the unit that all games are forced to use for their in-game lore, texts, etc. Games can be free to *pretend* that the node length is like a millimeter (to better fit the world's theme), even when in reality they still use the default size.

Wuzzy2 commented:
Why not call it a “node length” (n) instead of a meter? Or some other generic, neutral unit?
A “neter”, if you will. xD (OK, that last sentence wasn't serious.)

The word “meter” brings a lot of baggage and assumptions with it, and requires endless explanations to avoid confusion to make clear it's only meant as a dev shorthand and does not mean at all that games must always be in “human scale”, etc. Remember, goal of being generic!

If we just use a neutral unit, the endless explanations vanish with a poof. Because then it's obvious.

benrob0329 commented:
Using nodes as a form of in-game mesurement is not uncommon, for example describing speed as n/s rather than a raw velocity.

AKryukov92 commented:
There were some comments, but no explicit proposal about wording.
I propose to replace
> Games with roughly meter cubed cubes.

With
>* World consist of nodes and entities. They are smallest elements available for interaction.
* Nodes are staked in 3 dimensional grid on discrete positions. Distance between adjacent nodes is constant. Each position of the grid contains only 1 node.
* Distance between adjacent nodes is used as a base unit for measuring distances in game.

Wuzzy2 commented:
For clarity, I would also add:

> * Note: A node's physical or graphical appearance can differ from the standard 1×1×1 size and can even be larger (up to a factor of X) and overlap with other nodes

AKryukov92 commented:
There was an idea about worlds of different scale, which is relevant to this discussion.
https://forum.minetest.net/viewtopic.php?f=49&t=22602

Wuzzy2 commented:
> I have seen some guys on the forums, who use MT for educational purposes. I think it worth mentioning in mission too.

I disagree. Just because some random people use MT for this and that should not mean that Minetest has to follow suit.

MT is and should primarily remain a *game* engine, so the focus must lie on gameplay features. I'm afraid if you try to add non-game stuff into the direction document, you will water down the direction way too much and lose focus.

If it happens that people use MT for education, so be it, but this should only be a side effect, NOT the main focus of development.

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		Development will seek to improve the performance of rendering, to increase FPS on both old and new
+		devices. New graphical features may be added if they do not jeapordize performance when not used.
+* **Improve code structure and maintainability**
+		Development will seek to improve the maintainability of code through refactoring.
```
> Development will seek to improve the maintainability of code through refactoring.

Before you'll start refactoring, you should write TESTS. Auto tests are preferrable, but keeping suite of manual test-scenarios might be good for a start.

rubenwardy replied:
most of the code doesn't lend well to automated tests

nerzhul replied:
currently many parts of the code are too heavy to support tests, but refactored one should provide tests. We don't enforce them but we encourage when it's possible and we enforced when it's necessary

rubenwardy commented:
> I have seen some guys on the forums, who use MT for educational purposes. I think it worth mentioning in mission too.

I don't really know exactly what you mean by this, ie: exactly what you think should be added, but I don't think it's relevant to state such a specific usecase

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		Avoids relying on third party behaviour. For example, there is no
+		central account system and the serverlist isn't actually needed to play online.
+6. **Multi-platform**. Essential features should work on all platforms.
+7. **Gameplay over graphics**.
```
You could write something about "encouraging imagination" by graphics in MTG and default game.
Simplicity of MTG might be pretended to reduce entry threshold for new modders.
Detailisation of other games does not have to be primitive, but should be consistent.

AKryukov92 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		central account system and the serverlist isn't actually needed to play online.
+6. **Multi-platform**. Essential features should work on all platforms.
+7. **Gameplay over graphics**.
+8. **Maintain fork-friendliness**.
```
Is this something about "if team denies your PR and you made your own fork, then one day our team might merge your work to master"?

benrob0329 replied:
This line seems useless without explanation or context.
```suggestion
8. **Don't Be Hostile Towards Project Forks**.
		Attitude towards divergent code-bases and communities should remain friendly,
		and changes should not be made to spite these.
```

paramat replied:
It is better this line is stated in the positive "fork-friendliness" instead of the negative "do not be hostile", "... to spite these". The suggestion above is unpleasant and dramatic.
Besides that, this may not be what "fork-friendliness" is about.

benrob0329 replied:
You're probably right, a positive statement will sound better than a negative one. Perhaps Ruben can give some insight into what he meant.

nerzhul replied:
we must keep positive statement.
Also we should encourage backporting fixes & features as far it's possible

AKryukov92 commented:
What about implementing features related to collect data about players?
For example most crafted item, typical behaviour, client version, typical online hours and so on.

What about making money with using MT? I know that there is a license, but community might offend people trying to make money.

AKryukov92 commented:
I think there should be explicit meaning for words `game` and `mod` and how they relate to each-other.

I propose following:
> **Mod** is a packed set of lua scripts, sound and image assets which add some gameplay feature for player.
> **Game** is a standalone packed set of mods which provide consistent gameplay including world creation settings, player session settings, gameplay rules and player goals. Games are distributed via content distribution platform.
> Mods are developed as addition for some specific game. If not specified otherwise, they are developed for Minetest Game.

benrob0329 commented:
> Mods are developed as addition for some specific game. If not specified otherwise, they are developed for Minetest Game.

Mods can be generic, and can reasonably support multiple games assuming the games have similar crafting methods. MTG should not be given any extra status than it already has, and imo shouldn't even be shipped in default with the in-game ContentDB interface.

Wuzzy2 commented:
But we already have a definition of games and mods: https://dev.minetest.net/Terminology

AKryukov92 commented:
> Mods can be generic, and can reasonably support multiple games assuming the games have similar crafting methods.

Mods should not be generic. Every mod should target one and only one specific game. I have explained it [here](https://github.com/minetest/minetest/issues/8934#issuecomment-531732113). Generic mods are simple by themselves, but require enormous complexity of environment.

> MTG should not be given any extra status than it already has, and imo shouldn't even be shipped in default now that we have the in-game ContentDB interface.

It is reasonable to assume that mod without explicit hard dependency targets default game. So far MTG is only default game shipped by default. It means that mods target MTG by default.

sfan5 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+3. **Keep the engine lightweight**.
+		Support a large amount of gameplay by pushing code and functionality out
+		to Lua, rather than having a complex core. Keeping the core as small as
+		possible makes it easier to test and more flexible.
```
I'd like to have some balance mentioned here
since performance also matters, moving some features into the core <b>is</b> a good idea (even if lua could technically do it on its own)

nerzhul replied:
i don't agree with keeping engine lightweight, it's a wrong direction regarding the performance. I'd rather want to see generic more than lightweight. By generic i mean no explicit object, like firespread, chests or something like this.
But higher level Lua calls are always needed for the performance, and we may need also some frameworks to help minetest to become a game. We are not a game engine, we are mainly tied to voxel games, and we should offer features needed to do a finished an performant game for end users.

nerzhul replied:
The core complexity is currently due to a big code layering due to historic and non refactored code from the 0.3 era when minetest moved from a single game to a multi player game. We worked hard to improve it, but we have some layers which are currently crazy to maintain. 
For me, mapgen is one of the good candidates for a refactor (the mapblock generation process), but also the network layer, and the ServerEnvironment.

sfan5 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+  Nodes may exceed their bounds, but still belong to one position.
+* The visible size of a unit length may change according to the game's requirements, but the engine does
+  not aim to promise good performance for visible sizes far deviating the default.
+* Games will tend to define the unit length distance a meaning, for example, Minetest Game defines.
```
Minetest Game defines?

benrob0329 replied:
```suggestion
* Games will tend to define the unit length distance a meaning, for example, Minetest Game defines one node as one meter.
```

sfan5 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		and potential.
+* **Improved graphics and rendering performance**
+		Development will seek to improve the performance of rendering, to increase FPS on both old and new
+		devices. New graphical features may be added if they do not jeapordize performance when not used.
```
If this were true, shader would have been added.
this needs less absolute terms

DS-Minetest replied:
Also that word should probably be "je**o**p**a**rdize".

benrob0329 replied:
sfan5: Not entirely true. Shaders are fundamental to any non-ancient rendering pipeline, and can actually help performance when used correctly. MT does not use them correctly unfortunately.

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+* `Minetest` is the game engine.
+* `Minetest Game` is the default game which comes with the engine, and is not covered by this document.
+* The `Minetest Project` refers to the combination of the above, the development team, and supporting tools.
+
```
```suggestion
* `Content` refers to any installable add-ons, such as Mods, Games, and Texture Packs.

```

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+## 2. Aims
+
+1. **Stability and performance**,
+		especially on older devices.
```
```suggestion
		Mobile and low-end devices should not be excluded from essential graphical and gameplay elements.
```

nerzhul replied:
not older but low end yes.
I'm running my minetest server on rock 64 (which is currently the most powerful balanced with ram & cpu low end device server we can host at home), and minetest server should run fast on it.
For the client android is the good low end device target.
Note: currently running the most interesting mods on the rock 64 (and i think the RPI 4 too) cause minetest server to lag with 1 or 2 servers. I mean with item_drop which is just essential in survival farmining mod, and mob frameworks too (the get_obj_in_radius PR has improved the perf but we are very very far from disabling both mods)

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		If everything is either disableable or suitable for such, then it gives
+		much more freedom for people to experiment and create interesting things,
+		even in ground-based worlds.
+3. **Keep the engine lightweight**.
+		Support a large amount of gameplay by pushing code and functionality out
+		to Lua, rather than having a complex core. Keeping the core as small as
+		possible makes it easier to test and more flexible.
+4. **Prefer simplicity**.
+		Try to be simple in both API and implementing features.
```
```suggestion
3. **Keep The Core Simple**.
		Prefer simpler solutions and APIs; keeping the core as small as possible
		makes it easier to test and maintain. Pushing unneeded features to Lua
		allows mods and games to better implement custom behavior.
4. **Flexible APIs**.
		Support a wide variety of gameplay paradigms by making generic APIs
		that are easily configured or scripted in Lua.
```

nerzhul replied:
simple API != performance. We should have `coherent API` more than simple API. low level API means we will deny responsilibility and let modders do the same code again and again.
When some code it too repeated and can be integrated in a fashioned and performant way to the core, we should do it.

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+4. **Prefer simplicity**.
+		Try to be simple in both API and implementing features.
+5. **Self-contained**.
+		Avoids relying on third party behaviour. For example, there is no
```
```suggestion
		Avoid relying on third party behavior. For example, there is no
```

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		Try to be simple in both API and implementing features.
+5. **Self-contained**.
+		Avoids relying on third party behaviour. For example, there is no
+		central account system and the serverlist isn't actually needed to play online.
```
```suggestion
		central account system and the serverlist isn't needed to play online.
```

nerzhul replied:
i think we can say there will never be a central account system. We don't have the resources to manage that (in terms of time & money)

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		Avoids relying on third party behaviour. For example, there is no
+		central account system and the serverlist isn't actually needed to play online.
+6. **Multi-platform**. Essential features should work on all platforms.
+7. **Gameplay over graphics**.
```
```suggestion
7. **Flexible, Low-Tech Graphics**.
		Because of the need to support mobile and low-end devices, graphics should
		be suitably low-tech to maintain performance and consistency, however despite
		this the APIs given should be flexible enough that content can make the most
		of the engine's capabilities.
```

nerzhul replied:
i think we can say also that we can have nice graphics but it should be toggled by some options and not enabled by default for supporting various devices

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+
+* **Improving flexibility**.
+		To fulfill aim 2, development will seek to push functionality to Lua and increase the flexibility
+		and potential.
```
```suggestion
		To fulfill aim 2, development will seek to de-hardcode and increase the flexibility of the Lua API.
```

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+		Development will seek to improve the performance of rendering, to increase FPS on both old and new
+		devices. New graphical features may be added if they do not jeapordize performance when not used.
+* **Improve code structure and maintainability**
+		Development will seek to improve the maintainability of code through refactoring.
```
```suggestion
		Development will seek to improve the maintainability of code through refactoring and the addition of Unit Tests.
```

paramat commented on doc/direction.md:
```diff
@@ -0,0 +1,87 @@
+This document attempts to set out Minetest's philosophy, aims, and objectives.
+
+Please note that Minetest is developed by a community, and not everyone in the community
+holds the same views as to what Minetest should be.
```
This paragraph states the obvious so i think it is unnecessary. It also seems awkward and over-apologetic about these obvious and unavoidable differences.

nerzhul commented:
> Things also worth mentioning:
> 
>     * Something addressing why the discord patch was rejected. Maybe relating to free vs proprietary and also the low user benefit of it
> 
>     * More example gameplay targets

i'm a little bit late, but as our community seems to be discord centric, we should include the discord patch as an optional disabled function. If it doesn't cause problem with our license there is no problem to integrate it.

nerzhul commented:
i added my own review, but to be honest the minetest path is the same since i started to develop around 2014 and we improved code but we didn't changed anything in terms of adoption and fixes. We kept the craziness on the Lua flexiblity, which is good in terms of flex, but not good in terms of performance because we didn't added powerful C++ backed code to help mods to:
* write less redundant code across mods
* reduce their codebase because of engine feature lack
* increase the global performance in terms of cpu cycle and memory usage (don't forget Lua is an interpreted language this GC, on real time it's a problem not easy to handle, especially on the modders part, which are not full time developers which understand that).
On the refactor part, i may propose some design for reworking the network & mapgen pipelines, which can help increasing performance but also needs re-architecture

Wuzzy2 commented:
I'm all for making the engine more powerful to help mods reduce their own complexity.
Minetest's de facto path is to let mods do most of the “hard” work, and flexibilty is often the argument. But this argument flies out of the window when modders have to figure out the solutions to problems over and over again, but in the end it turns out there's really only one good solution anyway and we just end up with redundant code.
Yes, flexibility is a good thing, but it should not be an excuse for deliberately keeping the API too low-level, with no or very few tools to simplify common tasks.

rubenwardy commented:
Updated.

Question is - what should this document contain? Mission statement and aims are good, scope is questionable

I've removed the following parts:

> 7. **Gameplay over graphics**.
> 		To aid support of mobile and low-end devices, graphics should be
> 8. **Maintain fork-friendliness**

benrob0329 commented on doc/direction.md:
```diff
@@ -0,0 +1,91 @@
+		Support a large amount of gameplay by pushing code and functionality out
+		to Lua, where possible, rather than having a complex core.
+		The exception to this is when APIs are needed to increase performance.
+		Keeping the core as small as possible makes it easier to test and more flexible.
```
This section feels like it belongs in the sections preceding and following it. Keeping features in Lua sounds like a flexibility problem and not a performance one, and thus belongs in the sections referencing a gravitation-less space game. Similarly, stating that it's easier to maintain sounds like it belongs in the "Prefer simplicity" section as that's about code simplicity.

Basically, what is this section saying that isn't already, or couldn't just be stated in the other two sections?

benrob0329 commented:
> Question is - what should this document contain?

From what I've gathered, the purpose of this document is to reduce the bureaucracy by giving everyone a sane central ground to stand on, aka give the project a solid and indisputable direction to push efforts towards.

It should help answer questions like:
* Is this feature in scope?
* Is this in the spirit of the project?
* What are we aiming for, what should this project become?

Wuzzy2 commented:
> 3. **Keep the engine lightweight**.
>		Support a large amount of gameplay by pushing code and functionality out
>		to Lua, where possible, rather than having a complex core.
>		The exception to this is when APIs are needed to increase performance.
>		Keeping the core as small as possible makes it easier to test and more flexible.

I am not against the minimalism pholosophy in general, but I take great issue with extreme minimalism. Minetest has many features that aren't useful for ALL games, but they are still useful to a large number of games nonetheless.

I'm also afraid “lightweight” will be mistaken as an excuse to “dumb down the Lua API”.

I have seen so many times in PRs that core devs are blocking a new API functionality because it could be *technically* implemented in a game and therefore is “not neccessary”. Number 3 at least addresses performance, but it does not address simplification of game code.

If this aim is taken to its logical conclusion, we would have to delete many convenience functions that aren't technically needed, as they could be done in mods. We would have to ban e.g. `minetest.is_protected` or even health. Health actually has been proposed to be removed before, but this would be disastrous (simply because of all the dependencies).

If a new API feature can help to greatly reduce the code in various games but does not actually add anything to the game itself, it's still an improvement, but it would be in violation of aim 3. Yet getting rid of game code is a GOOD and important thing. If games are forced to repeat the same feature over and over again, by using the same algorithms over and over again, and using the same implementations over and over again, that's not good. That's a large amount of redundancy, and redundancy is bad.

Another reason for API functionality is interoperability. That would be `minetest.is_protected` again. If you force this feature to mods, it means that multiple mods have to compete with each other and my experience shows that this never ends up well, it's always a mess. It's much easier to have ONE core implementation that works well. That's why I love `minetest.is_protected` so much.

(Note that I don't care whether an API feature is implemented in C++ or builtin. I treat it equally.)

I caution about aggressively removing core features from Minetest with this aim as justification. If you want to force features to mods, the following things need to be true:

1) A mod implementation can perfectly replicate the old behavior and this has been proven
2) A mod implementation will not be significanty slower than the old behavior and this has been proven
3) There are different reasonable ways the given feature could be implemented. (But if there's only one reasonable implementation, that's an argument for keeping it in core intead.)

For ALL of the “remove health” or “remove breath” proposals, there is no evidence that 1 or 2 are true yet.


I propose that convenience functions be added as an explicit exception. With “convenience functions” I mean any function that doesn't add any “real” new functionality, but something that helps to solve a common task (a task shared by many games) with less code.

---

Apart from aim 3, I have no more complaints about this document.

paramat commented:
The intention of this PR is understandably a very difficult task, as rewriting and adding detail to celeron55's roadmaps generates large amounts of discussion, bikeshedding, disagreement and argument (see above), much of which is unresolvable as core devs have differing approaches to MT.
I can completely understand why rubenwardy might give up in this, i think anyone would, i think it is brave to attempt this.
> From what I've gathered, the purpose of this document is to reduce the bureaucracy by giving everyone a sane central ground to stand on, aka give the project a solid and indisputable direction to push efforts towards.

Celeron55's roadmaps already exist and are not much different to this PR, so this has already been done in a way.
An attempt to add details of direction not in celeron55's roadmaps will be very difficult to agree on by a large number of core devs who are equals.
So i think what you are hoping for cannot happen much more than it already happens.

The only way to have a detailed vision and direction (which is actually not necessary, as i have expressed elsewhere, the importance of this is hugely overstated) is if there is a dictator, but MT is a large number of core devs who are equals. Celeron55 is the only person who is in any kind of position to 'dictate', but decides to no longer do that.

So, i think all that can be done is to state what has been stated in celeron55's roadmaps. Even then, rewriting them means the new portrayal (which inevitably alters the meaning) is argued over.

As i suggested earlier, a very simple and low-effort way of doing this task is to just add celeron55's roadmaps and relevant writings to the repo as minimally-edited documents, this also stores them in a safe and suitable place. I would call this 'The Book Of Celeron55' =)
Much of the philosophy and spirit of MT is expressed in celeron55's sense of humour, laid-back attitude, sense of playfulness and of use of words, rewriting removes these very imporant aspects and creates a rather dry result with little sense of fun.
I do not object to a rewritten version, but the originals should also be present.

rubenwardy commented:
> The only way to have a detailed vision and direction is if there is a dictator

This isn't true, there's compromise and shared vision

Wuzzy2 commented:
There still has been no reply to my simplicity argument.

I actually already approved almost all of the document, just not this one section 3 (see above). I think it is very important to get this section right, and we hope we can agree on something.

It's not good if there are major disagreements about such a crucial aspect of development.

I'm just afraid the current sense of simplicity is being used to block some potentially very benefitial additions to the Lua API, because only simplicity in the core is considered, but not simplicity in the game code. But I wrote about in detail already (see above).

rubenwardy commented:
Closing in favour of #10536, this will be reopened anew after that PR is merged
